pkg/gcp/gcp_cloud_test.go

```diff
@@ -0,0 +1,206 @@
+/*
+SPDX-License-Identifier: Apache-2.0
+
+Copyright Contributors to the Submariner project.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package gcp_test
+
+import (
+	"errors"
+	"net/http"
+
+	"github.com/golang/mock/gomock"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/submariner-io/cloud-prepare/pkg/api"
+	"github.com/submariner-io/cloud-prepare/pkg/gcp"
+	"google.golang.org/api/compute/v1"
+	"google.golang.org/api/googleapi"
+)
+
+const ingressRuleName = "test-infraID-submariner-internal-ports-ingress"
+
+var _ = Describe("Cloud", func() {
+	Describe("PrepareForSubmariner", testPrepareForSubmariner)
+	Describe("CleanupAfterSubmariner", testCleanupAfterSubmariner)
+})
+
+func testPrepareForSubmariner() {
+	t := newCloudTestDriver()
+
+	var retError error
+
+	JustBeforeEach(func() {
+		retError = t.cloud.PrepareForSubmariner(api.PrepareForSubmarinerInput{
+			InternalPorts: []api.PortSpec{
+				{
+					Port:     100,
+					Protocol: "TCP",
+				},
+				{
+					Port:     200,
+					Protocol: "UDP",
+				},
+			},
+		}, api.NewLoggingReporter())
+	})
+
+	When("the firewall rule doesn't exist", func() {
+		BeforeEach(func() {
+			t.gcpClient.EXPECT().GetFirewallRule(projectID, ingressRuleName).Return(nil, &googleapi.Error{Code: http.StatusNotFound})
+		})
+
+		Context("", func() {
+			var actualRule *compute.Firewall
+
+			BeforeEach(func() {
+				t.gcpClient.EXPECT().InsertFirewallRule(projectID, gomock.Any()).DoAndReturn(func(_ string, rule *compute.Firewall) error {
+					actualRule = rule
+					return nil
+				})
+			})
+
+			It("should correctly insert it", func() {
+				Expect(retError).To(Succeed())
+
+				if actualRule != nil {
```

Comment:
This is set in the expected call to `InsertFirewallRule` on L70.  However if it's not actually called then it will fall through and let `gomock` fail it with their nice detailed message. I could check it here but it would be redundant.